File: task22_1.hs
-- Snakes and ladders
import Control.Monad.Reader
 
import Data.Set(Set)
import qualified Data.Set as Set
 
import Control.Applicative ((<$>))
import Data.Maybe (fromJust)
import Control.Monad.IO.Class
 
import Data.Array
import Data.List
 
data Square = Start | End | Roll | Go Int deriving (Show,Eq)
 
c2Square "S" = Start
c2Square "E" = End
c2Square "R" = Roll
c2Square  d  = Go $ read d
 
type Board a =  (ReaderT (Array Pos Square) IO) a
type Pos = Int
 
value :: Pos -> Board (Maybe Square)
value i = do
    array <- ask
    if inRange (bounds array) i then return (Just $ array!i)
    else return Nothing
 
go :: Pos -> Board (Set Pos)
go pos = go' <$> value pos
   where go' Nothing       = Set.empty
         go' (Just (Go i)) = Set.singleton (pos+i)
         go' _             = Set.fromList [pos+i | i<-[1..6]]
 
step :: Set Pos -> Board (Set Pos)
step positions = do
    new_poss <- mapM go $ Set.elems positions
    return $ Set.unions new_poss
 
containsEnd :: Set Pos -> Board Bool
containsEnd set = do
    values <- mapM value $ Set.elems set
    return $ or.map (==Just End) $ values
 
expand' :: Int -> Set Pos -> Set Pos -> Board (Maybe Int)
expand' turn last set = do
   --liftIO $ putStrLn.show $ set
   endReached <- containsEnd set
   if set `Set.isSubsetOf` last then return Nothing
   else if endReached then return $ Just turn
   else expand' (turn+1) (Set.union last set) =<< step set
 
expand :: Pos -> Board (Maybe Int)
expand start = expand' 0 Set.empty (Set.singleton start)
 
main :: IO ()
main = do
   n <- readLn :: IO Int
   lboard <- map c2Square <$> replicateM n getLine
   let start = (+1).fromJust $ findIndex (==Start) lboard
       array = listArray (1,n) lboard
   res <- runReaderT (expand start) array
   putStrLn $ maybe "impossible" show res

File: task22_1.js
// Snakes and ladders
// Read inputs from Standard Input (use readline()).
// Write outputs to Standard Output (use print()).

var n = readline();

var s = -1;
var e = -1;
var str = [];

for (var i = 0; i < n; i++) {
    var a = readline();
    if(a == 'S') {
        s = i;
    }
    if(a == 'E') {
        e = i;
    }
        
    str.push(a);
}


var q = [];
var visit = [];
q.push([s,0]);

var fin = 'impossible';

while(q.length) {
    var z = q[0][0];
    var step = q[0][1];
   
    q.splice(0,1);
    
    if(visit[z]) {
        continue;
    }
    
    //print(z+" "+step);
    visit[z] = true;
    
    if(str[z]=='E') {
        fin = step;
    } else if(str[z]=='R') {
         for(var i=1;i<=6;i++) {
            if(z + i <= str.length) {
                q.push([z+i,step+1]);
            }
        }       
    } else if(str[z]=='S') {
        for(var i=1;i<=6;i++) {
            if(z + i <= str.length) {
                q.push([z+i,step+1]);
            }
        }
    } else {
        var x = parseInt(str[z]);
        q.push([z+x,step+1]);
    }
}

print(fin);
